File: packages/go/qorechain/evm/precompiles.go
package evm

import (
	"fmt"
	"math/big"

	"github.com/qorechain/qorechain-sdk/packages/go/qorechain/address"
)

// Fixed 20-byte precompile addresses (zero-padded), matching the chain's
// published interface registrations. These mirror the canonical SDK constants.
const (
	// CrossVMBridgeAddress is the CrossVM Bridge precompile.
	CrossVMBridgeAddress = "0x0000000000000000000000000000000000000901"
	// PQCVerifyAddress is the PQC signature-verification precompile.
	PQCVerifyAddress = "0x0000000000000000000000000000000000000A01"
	// PQCKeyStatusAddress is the PQC key-registration-status precompile.
	PQCKeyStatusAddress = "0x0000000000000000000000000000000000000A02"
	// AIRiskScoreAddress is the AI transaction risk-score precompile
	// (IQoreAI.aiRiskScore).
	AIRiskScoreAddress = "0x0000000000000000000000000000000000000B01"
	// AIAnomalyCheckAddress is the AI anomaly-check precompile
	// (IQoreAI.aiAnomalyCheck).
	AIAnomalyCheckAddress = "0x0000000000000000000000000000000000000B02"
	// RLConsensusParamsAddress is the consensus-parameters precompile.
	RLConsensusParamsAddress = "0x0000000000000000000000000000000000000C01"
)

// Precompile function signatures (used to derive the 4-byte selectors).
const (
	aiRiskScoreSignature    = "aiRiskScore(bytes)"
	aiAnomalyCheckSignature = "aiAnomalyCheck(address,uint256)"
)

// AIRiskScore computes an on-chain risk score for raw transaction data by
// calling the aiRiskScore(bytes) precompile via eth_call.
//
// It returns the numeric score and a discrete level (0 = lowest risk). A level
// of 3 or above is conventionally treated as high risk (see
// SimulateWithRiskScore). A returned error usually means the precompile is not
// available on this node.
func (c *Client) AIRiskScore(txData []byte) (score *big.Int, level uint8, err error) {
	calldata := append(selector(aiRiskScoreSignature), encodeBytesArg(txData)...)
	ret, err := c.EthCall(CallMsg{To: normalizeHexAddr(AIRiskScoreAddress), Data: calldata}, "latest")
	if err != nil {
		return nil, 0, err
	}
	score, err = decodeUint(ret, 0)
	if err != nil {
		return nil, 0, fmt.Errorf("decode aiRiskScore score: %w", err)
	}
	level, err = decodeUint8(ret, 1)
	if err != nil {
		return nil, 0, fmt.Errorf("decode aiRiskScore level: %w", err)
	}
	return score, level, nil
}

// AIAnomalyCheck checks whether a (sender, amount) pair is anomalous by calling
// the aiAnomalyCheck(address,uint256) precompile via eth_call.
//
// sender may be an EVM "0x"-prefixed hex address or a bech32 ("qor1...")
// address; bech32 is converted to its 20-byte payload. It returns the anomaly
// score and a flagged boolean. A returned error usually means the precompile is
// not available on this node.
func (c *Client) AIAnomalyCheck(sender string, amount *big.Int) (anomalyScore *big.Int, flagged bool, err error) {
	hexSender, err := toHexAddress(sender)
	if err != nil {
		return nil, false, err
	}
	encAddr, err := encodeAddress(hexSender)
	if err != nil {
		return nil, false, err
	}
	calldata := append(selector(aiAnomalyCheckSignature), encAddr...)
	calldata = append(calldata, encodeUint(amount)...)
	ret, err := c.EthCall(CallMsg{To: normalizeHexAddr(AIAnomalyCheckAddress), Data: calldata}, "latest")
	if err != nil {
		return nil, false, err
	}
	anomalyScore, err = decodeUint(ret, 0)
	if err != nil {
		return nil, false, fmt.Errorf("decode aiAnomalyCheck score: %w", err)
	}
	flagged, err = decodeBool(ret, 1)
	if err != nil {
		return nil, false, fmt.Errorf("decode aiAnomalyCheck flagged: %w", err)
	}
	return anomalyScore, flagged, nil
}

// SimulateTx is the input to SimulateWithRiskScore.
type SimulateTx struct {
	// From is the sender address (hex "0x..." or bech32). Used as the eth_call
	// caller, the anomaly-check subject, and (when To is empty) for gas.
	From string
	// To is the target contract / recipient address.
	To string
	// Data is the transaction calldata; also fed to the risk-score precompile.
	Data []byte
	// Value is the transaction value in wei (nil = 0). Also the anomaly amount.
	Value *big.Int
}

// RiskResult holds the AI risk-score precompile output.
type RiskResult struct {
	Score *big.Int
	Level uint8
}

// AnomalyResult holds the AI anomaly-check precompile output.
type AnomalyResult struct {
	Score   *big.Int
	Flagged bool
}

// SimulateResult is the combined pre-flight result from SimulateWithRiskScore.
type SimulateResult struct {
	// Gas is the eth_estimateGas estimate for the transaction.
	Gas uint64
	// Risk is the aiRiskScore precompile output for the calldata.
	Risk RiskResult
	// Anomaly is the aiAnomalyCheck precompile output for (From, Value).
	Anomaly AnomalyResult
	// Safe is an ADVISORY verdict: true when Risk.Level < 3 AND not flagged.
	//
	// It is a convenience signal for client-side UX only and is NOT enforced by
	// the chain — the transaction can still be submitted when Safe is false, and
	// a Safe value of true is not a guarantee of execution success.
	Safe bool
}

// SimulateWithRiskScore runs an AI pre-flight over a prospective transaction: it
// estimates gas (eth_estimateGas), scores the calldata (aiRiskScore) and checks
// the sender/amount for anomalies (aiAnomalyCheck), then combines them into an
// advisory Safe verdict (Risk.Level < 3 && !Anomaly.Flagged).
//
// The Safe field is advisory only; see SimulateResult.Safe.
func (c *Client) SimulateWithRiskScore(tx SimulateTx) (SimulateResult, error) {
	value := tx.Value
	if value == nil {
		value = big.NewInt(0)
	}

	gas, err := c.EthEstimateGas(CallMsg{From: tx.From, To: tx.To, Data: tx.Data, Value: value})
	if err != nil {
		return SimulateResult{}, fmt.Errorf("estimate gas: %w", err)
	}

	score, level, err := c.AIRiskScore(tx.Data)
	if err != nil {
		return SimulateResult{}, fmt.Errorf("ai risk score: %w", err)
	}

	anomalyScore, flagged, err := c.AIAnomalyCheck(tx.From, value)
	if err != nil {
		return SimulateResult{}, fmt.Errorf("ai anomaly check: %w", err)
	}

	return SimulateResult{
		Gas:     gas,
		Risk:    RiskResult{Score: score, Level: level},
		Anomaly: AnomalyResult{Score: anomalyScore, Flagged: flagged},
		Safe:    level < 3 && !flagged,
	}, nil
}

// toHexAddress accepts either a "0x"-prefixed hex address (returned as-is) or a
// bech32 address (converted to its "0x" payload).
func toHexAddress(addr string) (string, error) {
	if len(addr) >= 2 && (addr[:2] == "0x" || addr[:2] == "0X") {
		return addr, nil
	}
	return address.Bech32ToHex(addr)
}

File: packages/py/src/qorsdk/precompiles.py
"""AI pre-flight bindings for QoreChain's EVM precompiles.

QoreChain's EVM Engine exposes on-chain risk/anomaly intelligence through two
fixed-address precompiles, callable like any view function via ``eth_call``:

- ``aiRiskScore(bytes) returns (uint256 score, uint8 level)`` at
  :data:`PRECOMPILE_AI_RISK_SCORE` — a model-driven risk score for raw
  transaction calldata.
- ``aiAnomalyCheck(address,uint256) returns (uint256 anomalyScore, bool flagged)``
  at :data:`PRECOMPILE_AI_ANOMALY_CHECK` — whether a ``(sender, amount)`` pair is
  anomalous.

This module ABI-encodes the calldata by hand (no web3 dependency): the 4-byte
selector is ``keccak256(signature)[:4]`` (reusing :func:`qorsdk.utils.keccak256`),
and the two return words are sliced from the 32-byte-aligned result. The helpers
take the EVM JSON-RPC client (:class:`~qorsdk.qor.QorClient`, which exposes
``eth_call``/``eth_estimateGas``) and are also surfaced as methods on it.

Availability note: on a default or community node these precompiles may return a
"not available" error; they are available on QoreChain network nodes. The
:func:`simulate_with_risk_score` ``safe`` verdict is ADVISORY — it summarizes the
advice (``level < 3`` and not ``flagged``) but does not gate or alter the
transaction; the caller decides whether to proceed.
"""

from __future__ import annotations

from typing import Any, Protocol

from .utils import keccak256

#: ``aiRiskScore(bytes)`` precompile address (20 bytes, zero-padded).
PRECOMPILE_AI_RISK_SCORE = "0x0000000000000000000000000000000000000B01"
#: ``aiAnomalyCheck(address,uint256)`` precompile address (20 bytes, zero-padded).
PRECOMPILE_AI_ANOMALY_CHECK = "0x0000000000000000000000000000000000000B02"

#: The exact Solidity function signatures the selectors are derived from.
_SIG_AI_RISK_SCORE = "aiRiskScore(bytes)"
_SIG_AI_ANOMALY_CHECK = "aiAnomalyCheck(address,uint256)"

_WORD = 32


class _EthCaller(Protocol):
    """The EVM JSON-RPC surface these helpers need (``QorClient`` satisfies it)."""

    def call(self, method: str, params: list[Any] | None = ...) -> Any: ...


def _strip_0x(value: str) -> str:
    return value[2:] if value.startswith(("0x", "0X")) else value


def selector(signature: str) -> bytes:
    """The 4-byte function selector: ``keccak256(signature)[:4]``."""
    return keccak256(signature)[:4]


def _encode_uint(value: int) -> bytes:
    """A non-negative integer as a 32-byte left-padded big-endian word."""
    if value < 0:
        raise ValueError("cannot ABI-encode a negative integer")
    return int(value).to_bytes(_WORD, "big")


def _encode_address(address: str) -> bytes:
    """A 20-byte hex address as a 32-byte left-padded word."""
    raw = bytes.fromhex(_strip_0x(address))
    if len(raw) != 20:
        raise ValueError(f"expected a 20-byte address, got {len(raw)} bytes")
    return raw.rjust(_WORD, b"\x00")


def _encode_bytes_arg(data: bytes) -> bytes:
    """ABI-encode a single dynamic ``bytes`` argument (head + tail).

    Layout: a 32-byte head offset (``0x20``), then the 32-byte length, then the
    payload right-padded to a 32-byte boundary.
    """
    length = _encode_uint(len(data))
    pad = (-len(data)) % _WORD
    body = data + b"\x00" * pad
    head_offset = _encode_uint(_WORD)
    return head_offset + length + body


def encode_ai_risk_score(tx_data: bytes) -> str:
    """Calldata for ``aiRiskScore(bytes)`` as a ``0x``-prefixed hex string."""
    return "0x" + (selector(_SIG_AI_RISK_SCORE) + _encode_bytes_arg(tx_data)).hex()


def encode_ai_anomaly_check(sender: str, amount: int) -> str:
    """Calldata for ``aiAnomalyCheck(address,uint256)`` as a hex string."""
    calldata = (
        selector(_SIG_AI_ANOMALY_CHECK)
        + _encode_address(sender)
        + _encode_uint(amount)
    )
    return "0x" + calldata.hex()


def _result_words(result: str, count: int) -> list[int]:
    """Slice an ``eth_call`` hex result into ``count`` 32-byte words as ints."""
    raw = bytes.fromhex(_strip_0x(result))
    if len(raw) < count * _WORD:
        raise ValueError(
            f"eth_call returned {len(raw)} bytes; expected at least "
            f"{count * _WORD} ({count} words)"
        )
    return [
        int.from_bytes(raw[i * _WORD : (i + 1) * _WORD], "big") for i in range(count)
    ]


def _eth_call(client: _EthCaller, to: str, data: str) -> str:
    result = client.call("eth_call", [{"to": to, "data": data}, "latest"])
    return result if isinstance(result, str) else "0x"


def ai_risk_score(client: _EthCaller, tx_data: bytes) -> dict[str, int]:
    """Compute an on-chain risk score for raw transaction calldata.

    :returns: ``{"score": int, "level": int}`` (``level`` is the low byte of the
        second return word).
    """
    result = _eth_call(client, PRECOMPILE_AI_RISK_SCORE, encode_ai_risk_score(tx_data))
    score, level = _result_words(result, 2)
    return {"score": score, "level": level & 0xFF}


def ai_anomaly_check(client: _EthCaller, sender: str, amount: int) -> dict[str, Any]:
    """Check whether a ``(sender, amount)`` pair is anomalous.

    :returns: ``{"anomaly_score": int, "flagged": bool}``.
    """
    result = _eth_call(
        client, PRECOMPILE_AI_ANOMALY_CHECK, encode_ai_anomaly_check(sender, amount)
    )
    anomaly_score, flagged = _result_words(result, 2)
    return {"anomaly_score": anomaly_score, "flagged": bool(flagged)}


def simulate_with_risk_score(client: _EthCaller, tx: dict[str, Any]) -> dict[str, Any]:
    """Pre-flight a transaction: estimate gas plus AI risk and anomaly checks.

    Runs three reads against the node:

    - ``eth_estimateGas`` for the transaction.
    - :func:`ai_risk_score` over the tx ``data`` (empty when absent).
    - :func:`ai_anomaly_check` over ``(from, value)``.

    :param tx: ``{"from", "to"?, "data"?, "value"?}``. ``value`` may be an int or
        a hex string; ``data`` is a ``0x``-prefixed hex string.
    :returns: ``{"gas": int, "risk": {...}, "anomaly": {...}, "safe": bool}``.

    The ``safe`` flag (``risk.level < 3`` and not ``anomaly.flagged``) is
    ADVISORY: it summarizes the AI advice and never blocks or mutates the tx.
    """
    sender = tx["from"]
    data_hex = tx.get("data") or "0x"
    value = tx.get("value", 0)
    amount = int(value, 16) if isinstance(value, str) else int(value)

    call_obj = {k: tx[k] for k in ("from", "to", "data", "value") if tx.get(k) is not None}
    gas_hex = client.call("eth_estimateGas", [call_obj])
    gas = int(gas_hex, 16) if isinstance(gas_hex, str) else int(gas_hex)

    tx_data = bytes.fromhex(_strip_0x(data_hex)) if data_hex != "0x" else b""
    risk = ai_risk_score(client, tx_data)
    anomaly = ai_anomaly_check(client, sender, amount)
    safe = risk["level"] < 3 and not anomaly["flagged"]
    return {"gas": gas, "risk": risk, "anomaly": anomaly, "safe": safe}


__all__ = [
    "PRECOMPILE_AI_RISK_SCORE",
    "PRECOMPILE_AI_ANOMALY_CHECK",
    "selector",
    "encode_ai_risk_score",
    "encode_ai_anomaly_check",
    "ai_risk_score",
    "ai_anomaly_check",
    "simulate_with_risk_score",
]
